### PR TITLE
Extract hash bugfixes

### PR DESCRIPTION
**Pull request by prestwich**
- add explicit length checks to extract hash to prevent unsafe usage
- remove extract script sig len

### CODE REVIEW

**Comment by sr-gi** [comment]
Since the length is extracted at the very beginning now, I think you can simplify the code a bit.

Something on the lines of:

```
 uint8 _len = uint8(_output[8]); 

 if (_len + 9 != _output.length || _len < 2) { 
    return hex""; 
 } 

 _len -= 2;
 uint8 _payloadLen = uint8(_output[10]);
 // Check for maliciously formatted witness outputs 
 if (_payloadLen != _len || (_payloadLen != 0x20 && _payloadLen != 0x14)) {
     return hex"";
 }

 return _output.slice(11, uint256(_len)); 
```

Not sure if the final cast to `uint256` is really necessary to call `slice`, just left it for consistency with the types you currently have.

Solidity-wise LGTM anyway, just being nitpicky.

**Comment by prestwich** [comment]
Made a few simplifications

**Comment by sr-gi** [comment]
LGTM

**Comment by prestwich** [comment]
@ErinHales this could use more in-depth review when you have time

File: testVectors.json [hunk 1]
```diff
@@ -640,60 +628,74 @@
...
-      "comment": "Maliciously formatted p2sh output",
-      "input": "0x000000000000000017a914FF",
+      "comment": "Maliciously formatted p2sh output (wrong last byte)",
+      "input": "0x000000000000000017a9140102030405060708090a0b0c0d0e0f101112131489",
       "jsError": "Maliciously formatted p2sh output",
       "golangError": "Maliciously formatted p2sh output",
       "rustError": "Maliciously formatted p2sh output"
     },
+    {
+      "comment": "Maliciously formatted p2sh output (payload reports as 0x15 by)",
```
Comment (ErinHales) [inline testVectors.json:687]: `0x15 bytes`

File: golang/btcspv/bitcoin_spv_test.go [hunk 2]
```diff
@@ -257,17 +257,6 @@ func (suite *UtilsSuite) TestExtractOutpoint() {
...
-	fixture := suite.Fixtures.ExtractOutputScriptLen
-
-	for i := range fixture {
-		testCase := fixture[i]
-		expected := testCase.Output
-		actual := btcspv.ExtractOutputScriptLen(testCase.Input)
-		suite.Equal(expected, actual)
-	}
-}
-
```
Comment (ErinHales) [inline golang/btcspv/bitcoin_spv_test.go:270]: Delete `ExtractOutputScriptLenTC` struct from `bitcoin_spv_test_types.go`

Comment (ErinHales) [reply golang/btcspv/bitcoin_spv_test.go:270]: As well as line 47 from `bitcoin_spv_test.go`:
```
ExtractOutputScriptLen       []tutils.ExtractOutputScriptLenTC     `json:"extractOutputScriptLen"`
```

File: solidity/contracts/BTCUtils.sol [hunk 3]
```diff
@@ -404,37 +404,49 @@ library BTCUtils {
...
                     return hex"";
                 }
                 return _output.slice(11, 20);
             }
         }
-        return hex"";  /* NB: will trigger on OPRETURN and non-standard that don't overrun */
+        return hex"";  /* NB: will trigger on OPRETURN and any non-standard that doesn't overrun */
     }
 
     /* ********** */
```
Comment (ErinHales) [inline solidity/contracts/BTCUtils.sol:452]: Delete `extractOutputScriptLen` from `BTCUtils.sol`, `BTCUtilsDelegate.sol` and `BTCUtilsTest.sol`.

**Pull request merged**